### DISCUSSION TITLE
Will cppfront detect all possible uninitialized variables?

### DISCUSSION THREAD

**Discussion by ntrel** [hsutter/cppfront#1048] [Q&A]
```c++
main: () =
{
    i := 0;
    p: std::unique_ptr<int>;
    while i < 3 next i++ {
        if !i { break; }
        p = new<int>;
    }
    std::cout << p* << "\n"; // no cppfront error
}
```
Due to the break, p seems to be uninitialized when it is dereferenced. This would also happen when the initial loop test is false.

BTW, on my system, I get a compile error from g++ and latest cppfront git.
g++ (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0

```
$ g++ --std=c++20 loopinit.cpp -I ~/git/cppfront/include/
In file included from loopinit.cpp:4:
/home/nick/git/cppfront/include/cpp2util.h: In instantiation of ‘void cpp2::impl::deferred_init<T>::construct(auto:61&& ...) [with auto:61 = {std::unique_ptr<int, std::default_delete<int> > (&)()}; T = std::unique_ptr<int>]’:
loopinit.cpp:23:20:   required from here
/home/nick/git/cppfront/include/cpp2util.h:774:76: error: no matching function for call to ‘std::unique_ptr<int>::unique_ptr(<brace-enclosed initializer list>)’
  774 | uct(auto&& ...args) -> void { cpp2_default.enforce(!init);  new (&data) T{CPP2_FORWARD(args)...};  init = true; }
      |                                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by bluetarpmedia** [comment]
In this case cppfront generates code that produces a runtime error, so to get the code to compile and run, change this line:
```cpp
p = new<int>;
```
to:
```cpp
p = new<int>(1);  // Or some value
```
And then you'll get a runtime "Contract violation" when trying to access `p`.
[Godbolt](https://cpp2.godbolt.org/z/cP1Mxd7Y7)

**Reply by ntrel** [reply]
OK, so tracking whether an initialization happened at runtime solves the problem. I also filed https://github.com/hsutter/cppfront/issues/1049.

**Comment by hsutter** [comment]
> ```c++
> main: () =
> {
>     i := 0;
>     p: std::unique_ptr<int>;
>     while i < 3 next i++ {
>         if !i { break; }
>         p = new<int>;
>     }
>     std::cout << p* << "\n"; // no cppfront error
> }
> ```
>   
> Due to the break, p seems to be uninitialized when it is dereferenced. 

This is a bug in cppfront, because initialization should not be allowed in a loop (the break shouldn't affect that). Thanks for pointing this out!

**Comment by hsutter** [comment]
Moved to bug #1133
